src/api/OnmsHTTPOptions.ts

```diff
@@ -1,118 +1,338 @@
+        } else {
+          this._parameters[parameter] = String(v);
+        }
+      }
+    }
+    return this;
+  }
+
+  /**
+   * The data to use in the request.
+   *
+   * If `undefined` is passed, the data will be cleared.
+   * @param data the data
+   */
+  public data(data?: any) {
+    this._data = data;
+    return this;
+  }
+}
+// tslint:enable:completed-docs variable-name whitespace
+
 /**
  * Options to be used when making HTTP ReST calls.
  * @module OnmsHTTPOptions
  */
 export class OnmsHTTPOptions {
-  /** How long to wait for ReST calls to time out. */
-  public get timeout(): number {
-    if (this[TIMEOUT_PROP]) {
-      return this[TIMEOUT_PROP];
-    }
-    return DEFAULT_TIMEOUT;
+  /**
+   * Create a new builder for an [[OnmsHTTPOptions]] object.
+   * @param options if an existing options object is passed, the builder will be pre-populated
+   */
+  public static newBuilder(options?: OnmsHTTPOptions) {
+    return new OnmsHTTPOptionsBuilder(options);
   }
 
-  public set timeout(t: number) {
-    this[TIMEOUT_PROP] = t;
+  /** How long to wait for ReST calls to time out. */
+  public get timeout(): number | undefined {
+    return this[TIMEOUT_PROP] || undefined;
   }
 
   /** The authentication config that should be used. */
-  public get auth(): OnmsAuthConfig {
-    if (this[AUTH_PROP]) {
-      return this[AUTH_PROP];
+  public get auth(): OnmsAuthConfig | undefined {
+    const auth = this[AUTH_PROP];
+    if (auth !== null && auth !== undefined) {
+      return auth;
     }
     if (this.server && this.server.auth) {
       return this.server.auth;
     }
-    return {} as OnmsAuthConfig;
-  }
-
-  public set auth(a: OnmsAuthConfig) {
-    this[AUTH_PROP] = a;
+    return undefined;
   }
 
-  /** The server to use if no server is set on the HTTP implementation. */
-  public server?: OnmsServer | null;
+  /** The server to use instead of that provided by the HTTP implementation. */
+  public readonly server: OnmsServer | null;
 
   /** HTTP headers to be passed to the request. */
-  public headers = {} as IHash<string>;
+  public readonly headers = {} as IHash<string>;
 
   /** HTTP parameters to be passed on the URL. */
-  public parameters = {} as IHash<string | string[]>;
+  public readonly parameters = {} as IHash<string | string[]>;
 
   /** HTTP data to be passed when POSTing */
-  public data: any;
+  public readonly data: any;
 
   /**
    * The default timeout associated with these options.
    *
    * This is a trick for making sure serialization to JSON happens properly
    * without exposing internals.
    */
-  private [TIMEOUT_PROP]: number;
+  private readonly [TIMEOUT_PROP]: number | null;
 
   /**
    * The default authentication credentials associated with these options.
    *
    * This is a trick for making sure serialization to JSON happens properly
    * without exposing internals.
    */
-  private [AUTH_PROP]: OnmsAuthConfig;
+  private [AUTH_PROP]: OnmsAuthConfig | null;
```

Comment:
Looks like this can be readonly too. Need to update the constructor to only set it once.
```suggestion
  private readonly [AUTH_PROP]: OnmsAuthConfig | null;
```